pkg/build/strategies/external/external.go

```diff
@@ -0,0 +1,154 @@
+// wll determine external builder name, and api.Config feeds command's template variables. It can
+// return error in case of template parsing or evaluation issues.
+func (e *External) renderCommand(config *api.Config) (string, error) {
+	commandTemplate, exists := commands[config.WithBuilder]
+	if !exists {
+		return "", fmt.Errorf("cannot find command '%s' in dictionary: '%#v'",
+			config.WithBuilder, commands)
+	}
+
+	t, err := template.New("external-command").Parse(commandTemplate)
+	if err != nil {
+		return "", err
+	}
+	var output bytes.Buffer
+	if err = t.Execute(&output, config); err != nil {
+		return "", err
+	}
+	return output.String(), nil
+}
+
+// execute the given external command using "os/exec". Returns the outcomes as api.Result, making
+// sure it only marks result as success when exit-code is zero. Therefore, it returns errors based
+// in external command errors, so "s2i build" also fails.
+func (e *External) execute(externalCommand string) (*api.Result, error) {
+	log.V(0).Infof("Executing external build command: '%s'", externalCommand)
+
+	externalCommandSlice := strings.Split(externalCommand, " ")
+	cmd := exec.Command(externalCommandSlice[0], externalCommandSlice[1:]...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	res := &api.Result{Success: false}
+	res.Messages = append(res.Messages, fmt.Sprintf("Running command: '%s'", externalCommand))
+	err := cmd.Start()
+	if err != nil {
+		res.Messages = append(res.Messages, err.Error())
+		return res, err
+	}
+
+	if err := cmd.Wait(); err != nil {
+		if exitErr, okay := err.(*exec.ExitError); okay {
+			if status, okay := exitErr.Sys().(syscall.WaitStatus); okay {
+				exitCode := status.ExitStatus()
+				log.V(0).Infof("External command return-code: %d", exitCode)
+				res.Messages = append(res.Messages, fmt.Sprintf("exit-code: %d", exitCode))
+				if exitCode == 0 {
+					res.Success = true
+				} else {
+					return res, exitErr
+				}
+			}
+		} else {
+			return res, err
+		}
+	}
+	return res, nil
+}
+
+// asDockerfile inspect config, if user has already informed `--as-dockerfile` option, that's simply
+// returned, otherwise, considering `--working-dir` option first before using artificial name.
+func (e *External) asDockerfile(config *api.Config) string {
+	if len(config.AsDockerfile) > 0 {
+		return config.AsDockerfile
+	}
+
+	if len(config.WorkingDir) > 0 {
+		return path.Join(config.WorkingDir, s2iDockerfile)
+	}
+	return s2iDockerfile
+}
+
+// Build triggers the build of a "strategy/dockerfile" to obtain "AsDockerfile" first, and then
+// proceed to execute the external command.
+func (e *External) Build(config *api.Config) (*api.Result, error) {
+	config.AsDockerfile = e.asDockerfile(config)
+
+	externalCommand, err := e.renderCommand(config)
+	if err != nil {
+		return nil, err
+	}
+
+	// generating dockerfile following AsDockerfile directive
+	err = e.dockerfile.CreateDockerfile(config)
```

Comment:
> The more I think about it, why are we not using e.dockerfile.Build here? 

@adambkaplan using `e.dockerfile.Build` would require Docker, right? The idea is to have a sidekick of `--as-dockerfile`.

> There is a whole host of use cases which occur to assemble the dockerfile (script download, injections, etc.). It may mean that all builds run in a temporary directory, and not in the user's current directory containing the source code.

Okay, understood. In this PR we are using [`config.WorkingDir`](https://github.com/otaviof/source-to-image/blob/f75c941c92b0c02e98ff794c6be1ff77b6c762eb/pkg/build/strategies/external/external.go#L33) as the context directory, would that be applicable considering the location of the source-code and temporary directory?

Reply:
@otaviof & @adambkaplan could we resolve this, please, soon? :) We should be taking this PR in, and making it part of the productized image.

Reply:
@otaviof `e.dockerfile.Build` does not require a docker socket - that flow is invoked in OCP 4 builds.

Reply:
@adambkaplan check, thanks for the information. I think for the feature intended on this PR we can use only `CreateDockerfile`, right?

Reply:
@adambkaplan indeed, you're correct on this. Changing the implementation to build `Build` instead. Thanks!